### ISSUE TITLE
generate_image_left

### ISSUE THREAD

**Issue by Yelen719** [mrharicot/monodepth#67]
Hi Clément,

Thanks a lot for sharing the code! It works very well!

I'm a bit confused about your approach of generate_image_left/right:
As far as I see,  model.disp_left_est can be seen as a (scaled) inverse depth of the **left image**, but why do you perform self.generate_image_left(self.right_pyramid[i], self.disp_left_est[i]), which means you use model.disp_left_est on **right image**, to generate the synthetic left image? Probably I'm wrong, but as far as I understand, the **inverse depth of left image**(model.disp_left_est) should be applied to **the left image** to generate **the right image**. But your approach is to use the **inverse depth of left image** on **the right image** to generate **the right image**. 

Could you give me some clue about this?

Thanks a lot!
Yelen

**Comment by mrharicot** [comment]
Hi Yelen,

Did you read [our paper](https://arxiv.org/pdf/1609.03677.pdf)?
We explain this in Section 3 and Figure 3.

In order to generate a new image you have two options: forward and backward mapping.
You can see disparities as texture coordinates offsets:
in forward mapping: disparity == **where each pixel goes in the other image**
in backward mapping: disparity == **where does each pixel comes from in the other image**
While forward mapping might be easier to understand, it is rarely used at it will likely produce holes in the resulting image, and is not differentiable.

I hope this helps!

**Comment by Yelen719** [comment]
Hi Clement,

Thanks a lot for your prompt reply! I didn't know backward mapping before. This really helps a lot.

I notice that there is an issue about the potential bug of the calculation of disparity smoothness, I modified the code and it works, I hope this also helps you:

```
    def gradient_x(self, img):
        gx = img[:, :, :-1, :] - img[:, :, 1:, :]
        fill_zero = tf.zeros([tf.shape(gx)[0], tf.shape(gx)[1], 1, tf.shape(gx)[3]])
        gx = tf.concat([gx, fill_zero], axis=2)
        return gx

    def gradient_y(self, img):
        gy = img[:, :-1, :, :] - img[:, 1:, :, :]
        fill_zero = tf.zeros([tf.shape(gy)[0], 1, tf.shape(gy)[2], tf.shape(gy)[3]])
        gy = tf.concat([gy, fill_zero], axis=1)
        return gy
```

```
    def get_disparity_smoothness(self, disp, pyramid):
        disp_gradients_x = [self.gradient_x(d) for d in disp]
        disp_gradients_y = [self.gradient_y(d) for d in disp]

        image_gradients_x = [self.gradient_x(img[:,:,:,0:3]) for img in pyramid]
        image_gradients_y = [self.gradient_y(img[:,:,:,0:3]) for img in pyramid]

        weights_x = [tf.exp(-tf.reduce_mean(tf.abs(g), 3, keep_dims=True)) for g in image_gradients_x]
        weights_y = [tf.exp(-tf.reduce_mean(tf.abs(g), 3, keep_dims=True)) for g in image_gradients_y]

        smoothness_x = [tf.abs(disp_gradients_x[i]) * weights_x[i] for i in range(4)]
        smoothness_y = [tf.abs(disp_gradients_y[i]) * weights_y[i] for i in range(4)]
        return map(tf.add, smoothness_x, smoothness_y)
```
BTW, did you try to use other kernel like Sobel operator to calculate the gradient?

Thanks a lot!

**Comment by mrharicot** [comment]
Thanks, there is indeed a bug in the smoothness, I suspect it won't not really change the results.
I do plan to clean the code and fix bugs at some point.

I did not try using other gradient operators.
L1 smoothness is probably not the best choice either.
